### PR TITLE
[fbw_gps] GPS in FBW/FTD (e.g. geofencing)

### PR DESCRIPTION
**Pull request by dewagter**

### CODE REVIEW

File: sw/airborne/subsystems/intermcu/intermcu_ap.c [hunk 1]
```diff
@@ -144,6 +164,56 @@ static inline void intermcu_parse_msg(void (*rc_frame_handler)(void))
       break;
     }
 
+#if TELEMETRY_INTERMCU
```
Comment (flixr) [inline sw/airborne/subsystems/intermcu/intermcu_ap.c]: Seems like this should be part of another PR...

File: sw/airborne/subsystems/abi_sender_ids.h [hunk 2]
```diff
@@ -162,12 +162,16 @@
 #define GPS_SIM_ID 11
 #endif
 
+#ifndef GPS_IMCU_ID
```
Comment (flixr) [inline sw/airborne/subsystems/abi_sender_ids.h]: Would be better to add a new ID 14 instead of changing the IDs of existing ones...

File: sw/airborne/subsystems/intermcu/intermcu_ap.c [hunk 3]
```diff
@@ -144,6 +164,56 @@ static inline void intermcu_parse_msg(void (*rc_frame_handler)(void))
...
+      uint8_t size = DL_IMCU_DATALINK_msg_length(imcu_msg_buf);
+      uint8_t *msg = DL_IMCU_DATALINK_msg(imcu_msg_buf);
+      telemetry_intermcu_on_msg(0, msg, size);
+      break;
+    }
+
+#endif
+
+#if IMCU_GPS
+    case DL_IMCU_REMOTE_GPS: {
```
Comment (flixr) [inline sw/airborne/subsystems/intermcu/intermcu_ap.c]: Is still missing in pprzlink...

File: sw/airborne/subsystems/intermcu/intermcu_fbw.c [hunk 4]
```diff
@@ -66,10 +66,21 @@ static void intermcu_parse_msg(void (*commands_frame_handler)(void));
...
+static abi_event gps_ev;
+static void gps_cb(uint8_t sender_id, uint32_t stamp, struct GpsState *gps_s);
+#endif
+
 void intermcu_init(void)
 {
   pprz_transport_init(&intermcu.transport);
 
+#if USE_GPS
+  AbiBindMsgGPS(ABI_BROADCAST, &gps_ev, gps_cb);
```
Comment (flixr) [inline sw/airborne/subsystems/intermcu/intermcu_fbw.c]: The ABI ID used here should probably be configurable?

**Comment by podhrmic** [comment]
Can you describe how this fits together with existing geofencing features we have in the flight plan? See #1802

**Comment by dewagter** [comment]
@podhrmic Currently FBW does not run any flightplan, so GeoFence must be done in a module.

**Comment by dewagter** [comment]
- [x] GPS driver in FBW
 - [ ] use GPS in FBW
 - [ ] pprzlink update
 - [ ] forward FBW GPS to AP

**Comment by dewagter** [comment]
PPRZLINK update pending: https://github.com/paparazzi/pprzlink/pull/34

**Comment by dewagter** [comment]
Ready

File: sw/airborne/subsystems/intermcu/intermcu_fbw.c (hunk 4 above)
Comment (flixr) [inline sw/airborne/subsystems/intermcu/intermcu_fbw.c]: Think this should still be changed from ABI_BROADCAST to something like `IMCU_GPS_ID` which should default to `GPS_MULTI_ID`
Then you could also have multiple GPS on the fbw controller...

Comment (flixr) [inline sw/airborne/subsystems/intermcu/intermcu_fbw.c]: And using ABI_BROADCAST here, gps_cb gets actually called twice per new GPS message: once from the actual GPS implementation and once from gps.c with GPS_MULTI_ID

**Comment by flixr** [comment]
Also what about the other gps modules, right now only ublox can be used on fbw...

**Comment by gautierhattenberger** [comment]
When #1835 is merged, we'll have the possibility to group modules functions into groups. Maybe it will help to call only a subpart of the modules in FBW, so all GPS modules would be available in this case.

**Comment by dewagter** [comment]
- [x] ICMU_GPS_ID
 
Are you sure you still want me to convert ALL gps modules to run on FBW? Or await #1835?

**Comment by flixr** [comment]
I don't see how #1835 is related to that...
In any case it would make much more sense to change the target for gps modules from only `ap` to `!nps|sim`, then they are usable for any target except simulation...

**Comment by gautierhattenberger** [comment]
The PR itself is not related, but I was think that the splitting into tasks could help here. After looking again at the code, you're right, it doesn't.

**Comment by flixr** [comment]
To be done here:
- resolve conflicts
- update of module xml of other GPS modules

**Comment by dewagter** [comment]
- [x] conflicts resolved
 - [x] other GPS modules
